source/io/module_javaScriptInvoke.js

```diff
@@ -111,17 +200,35 @@ var assignJavaScriptInvoke;
          */
         Module.javaScriptInvoke.writeJavaScriptRefNum = function (javaScriptValueRef, jsValue) {
             var isStaticReference = Module.typeHelpers.isJSObjectStaticRefnum(javaScriptValueRef.typeRef);
-            if (isStaticReference) { // static reference (ie, control reference) shares cookie for the same jsValue
-                var cachedCookie = getCachedRefNumCookie(jsValue);
-                if (cachedCookie !== undefined) {
-                    Module.eggShell.writeDouble(javaScriptValueRef, cachedCookie); // set the VIA local to be this cookie value
-                    return;
-                }
+            var cookie;
+            if (isStaticReference) {
+                cookie = staticRefnumManager.set(jsValue);
+                Module.eggShell.writeDouble(javaScriptValueRef, cookie);
+                return;
             }
 
-            var newCookie = generateUniqueRefNumCookie();
-            Module.eggShell.writeDouble(javaScriptValueRef, newCookie);
-            cacheRefNum(newCookie, jsValue);
+            cookie = dynamicRefnumManager.set(jsValue);
+            Module.eggShell.writeDouble(javaScriptValueRef, cookie);
+        };
+
+        Module.javaScriptInvoke.isJavaScriptRefNumValid = function (javaScriptValueRef) {
+            var cookie = Module.eggShell.readDouble(javaScriptValueRef);
+            var isStaticReference = Module.typeHelpers.isJSObjectStaticRefnum(javaScriptValueRef.typeRef);
+            if (isStaticReference) {
+                return staticRefnumManager.has(cookie);
+            }
+            return dynamicRefnumManager.has(cookie);
+        };
+
+        // static reference (static control reference) always returns true (the operation was successful, which for static references is a no-op)
+        // dynamic reference (JS opaque reference from JSLI) returns true if the reference was found + cleared and false if the reference was invalid
+        Module.javaScriptInvoke.clearJavaScriptRefNum = function (javaScriptValueRef) {
+            var isStaticReference = Module.typeHelpers.isJSObjectStaticRefnum(javaScriptValueRef.typeRef);
+            if (isStaticReference) {
+                return true;
```

Comment:
Why doesn't this case have to clear the `_cookieToJsValueMap`?

Reply:
Static references are not closed by Close Reference; the diagram cannot close static references. It's a no-op. Maybe in the future we can add a boolean to force close or something to the public api but there hasn't been a need yet.

Reply:
I'll add a comment for that